### PR TITLE
Recoil Rework [RTM]

### PR DESCRIPTION
**Pull request by Humonitarian**
## About The Pull Request

Datumizes recoil, reworks some components, improves code quality, stops roaches from unnecessarily proccing recoil.

Recoil has been fundamentally changed in multiple ways, with one example being wielding handguns now does help reduce their kickback.

The bigger the gun, the less accurate it is with one hand, but more accurate it becomes when wielded or braced.

Total recoil is based on the projectile (.40 has more recoil than .35, etc.), and the gun's length (smg, carbine, lmg, etc.), as well as a per-gun basis number that will ensure the intended use of every weapon (one handed handguns, two handed longarms) will keep the old ratios. This number can be effortlessly tweaked to balance the guns later on.

Attachment with the exception of the gyro stabilizer will now modify all penalties (shooting with one hand, etc.), not just base recoil.

## Why It's Good For The Game

More dynamic, more modular, cleaner code, bugfixes.

## Changelog
:cl:
refactor: recoil within guns refactored
refactor: bullets determine base recoil
tweak: you can now wield every gun to reduce recoil, including pistols. Bracing also reduces recoil. The reduction of both scales with gun size
fix: roaches no longer suffer from recoil
balance: recoil was tweaked to feel closer to the october values, giving FA a slight underhand
/:cl:

### CODE REVIEW

**Comment by MLGTASTICa** [comment]
Surprise surprise , YET another major combat rebalance.

**Comment by TheShown911** [comment]
Based on armor datum? Cringe!

**Comment by SirRichardFrancis** [comment]
Major combat rebalances are just like NT updates. Nobody asks for them, yet they keep coming and coming.

**Comment by Humonitarian** [comment]
> Major combat rebalances are just like NT updates. Nobody asks for them, yet they keep coming and coming.

Goal isn't to make it a major combat rebalance.
Much like the original dynamic recoil dissipation, the user-end should feel mostly only QoL changes. Value tweaks are planned to move recoil to their old values, similarly to the fire-delay based recoil adjustment.

**Comment by Alliostra** [comment]
Firstly there is absolutely no reason for this to be made, secondly it will screw up recoil and make all these guns impossible to shoot, thirdly, one and possibly more gun recoil changes are direct violation to contributing guidelines, you should try reading them.
https://wiki.cev-eris.com/Content_GuidelinesEn

**Comment by Iglpigl** [comment]
Firstly there is absolutely no reason for this to be made, secondly it will screw up recoil and make all these guns impossible to shoot, thirdly, one and possibly more gun recoil changes are direct violation to contributing guidelines, you should try reading them.
https://wiki.cev-eris.com/Content_GuidelinesEn

**Comment by ValoTheValo** [comment]
Making Recoil based on weapon type isnt a good idea. it makes less modularity and ability for weapons to feel distinct than before. before you could just edit one number now you have to crank out a whole preset from what i understand

**Comment by ValoTheValo** [comment]
and tying recoil to damage is just honestly kind of dumb and janky.

File: code/modules/mob/living/carbon/human/human_recoil.dm [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+	if(ishuman(src))
+		var/mob/living/carbon/human/H = src
+		var/suit_stiffness = 0
+		var/uniform_stiffness = 0
+		if(H.wear_suit)
+			suit_stiffness = H.wear_suit.stiffness
+		if(H.w_uniform)
+			uniform_stiffness = H.w_uniform.stiffness
+		base_recoil += suit_stiffness + suit_stiffness * uniform_stiffness // Wearing it under actual armor, or anything too thick is extremely uncomfortable.
+	add_recoil(base_recoil)
```
Comment (SirRichardFrancis) [inline code/modules/mob/living/carbon/human/human_recoil.dm:15]:
```suggestion
/mob/living/carbon/human/handle_movement_recoil()
	deltimer(recoil_reduction_timer)
	total_recoil = 1 + H.wear_suit.stiffness? + H.wear_suit.stiffness? * H.w_uniform.stiffness? // Wearing it under actual armor, or anything too thick is extremely uncomfortable.
	add_recoil(total_recoil)
```
It's `/human` proc, hence neither type check nor type assignment is needed.

File: code/modules/mob/living/living_recoil.dm [hunk 2]
```diff
@@ -1,64 +1,36 @@
-/mob/living/proc/handle_recoil(var/obj/item/gun/G)
+/mob/living/proc/handle_recoil(var/obj/item/gun/G, var/recoil_buildup)
```
Comment (SirRichardFrancis) [inline code/modules/mob/living/living_recoil.dm:1]:
```suggestion
/mob/living/proc/handle_recoil(obj/item/gun/G, recoil_buildup)
```

File: code/modules/mob/living/living_recoil.dm [hunk 3]
```diff
@@ -1,64 +1,36 @@
...
-		recoil += G.brace_penalty
 
 	var/debug_recoil = min(0.3, G.fire_delay)
 	if(G.fire_delay == 0)
 		debug_recoil = 0.3
 	if(G.burst > 1)
 		debug_recoil = max(debug_recoil, G.burst_delay)
 
-	add_recoil(G.recoil_buildup + debug_recoil) // DEBUG, remove the debug_recoil after testing is over!
+	add_recoil(recoil_buildup + debug_recoil) // DEBUG, remove the debug_recoil after testing is over!
```
Comment (SirRichardFrancis) [inline code/modules/mob/living/living_recoil.dm:10]: image

File: code/modules/projectiles/gun.dm [hunk 4]
```diff
@@ -40,11 +40,19 @@
...
 	var/fire_sound = 'sound/weapons/Gunshot.ogg'
 	var/rigged = FALSE
-	var/braced = FALSE //for gun_brace proc.
-	var/brace_penalty = FALSE//penalty if not braced.
-	var/braceable = 1 //can the gun be used for gun_brace proc, modifies recoil. If the gun has foregrip mod installed, it's not braceable. Bipod mod increases value by 1.
 	var/fire_sound_text = "gunshot"
+
+	var/datum/recoil/recoil // Reference to the recoil datum in datum/recoil.dm
+
+	//TODO: remove this
```
Comment (SirRichardFrancis) [inline code/modules/projectiles/gun.dm:47]: image

File: code/modules/projectiles/gun.dm [hunk 5]
```diff
@@ -452,23 +465,49 @@
...
+			if(1 to 2)
+				if(prob(25)) //don't need to tell them every single time
+					to_chat(user, "<span class='warning'>Your aim wavers slightly.</span>")
+			if(2 to 3)
+				if(prob(50)) //don't need to tell them every single time
+					to_chat(user, "<span class='warning'>Your aim wavers as you fire \the [src] while carrying it.</span>")
+			if(3 to 5)
+				to_chat(user, "<span class='warning'>You have trouble keeping \the [src] on target while carrying it!</span>")
+			if(5 to INFINITY)
+				to_chat(user, "<span class='warning'>You struggle to keep \the [src] on target while carrying it!</span>")
```
Comment (SirRichardFrancis) [inline code/modules/projectiles/gun.dm:507]:
```suggestion
	if(unwielded_recoil)
		switch(unwielded_recoil)
			if(1 to 2)
				if(prob(25)) // Don't need to tell them every single time
					to_chat(user, SPAN_WARNING("Your aim wavers slightly."))
			if(2 to 4)
				if(prob(50))
					to_chat(user, SPAN_WARNING("Your aim wavers as you fire \the [src] with just one hand."))
			if(4 to 7)
				to_chat(user, SPAN_WARNING("You have trouble keeping \the [src] on target with just one hand."))
			if(7 to INFINITY)
				to_chat(user, SPAN_WARNING("You struggle to keep \the [src] on target with just one hand!"))

	else if(brace_recoil)
		switch(brace_recoil)
			if(1 to 2)
				if(prob(25))
					to_chat(user, SPAN_WARNING("Your aim wavers slightly."))
			if(2 to 3)
				if(prob(50))
					to_chat(user, SPAN_WARNING("Your aim wavers as you fire \the [src] while carrying it."))
			if(3 to 5)
				to_chat(user, SPAN_WARNING("You have trouble keeping \the [src] on target while carrying it!"))
			if(5 to INFINITY)
				to_chat(user, SPAN_WARNING("You struggle to keep \the [src] on target while carrying it!"))
```

File: code/modules/projectiles/projectile/beams.dm [hunk 6]
```diff
@@ -12,6 +12,7 @@
 	hitscan = 1
 	invisibility = 101	//beam projectiles are invisible as they are rendered by the effect engine
 	style_damage = 30 //hitscan, light speed projectiles? Be glad its easier to dodge than a revolver.
+	recoil = 1 // Even less than self-propelled bullets
```
Comment (SirRichardFrancis) [inline code/modules/projectiles/projectile/beams.dm:15]: Recoil on lasers? C'mon.

File: code/modules/projectiles/projectile/energy.dm [hunk 7]
```diff
@@ -82,19 +85,21 @@
...
+	recoil = 5
 
 /obj/item/projectile/energy/neurotoxin
 	name = "neuro"
 	icon_state = "neurotoxin"
 	damage_types = list(TOX = 5)
 	weaken = 5
 
-/obj/item/projectile/energy/plasma
+/obj/item/projectile/energy/plasma // What?
```
Comment (SirRichardFrancis) [inline code/modules/projectiles/projectile/energy.dm:102]: ?

**Review by SirRichardFrancis — CHANGES_REQUESTED** [review]
Reworking how the guns work is what our codebase need, but frankly, I'm not fan of changes this PR brings.
Not that they're bad necessarily, just quite pointless and screwing balance in unpredictable way.
Too much needless math for every single shot, and moving existing vars into separate datums for I dunno what gain.

File: code/__DEFINES/weapons.dm [hunk 8]
```diff
@@ -75,3 +75,11 @@
 #define RECOIL_REDUCTION_TIME 1 SECOND
 
 #define VIG_OVERCHARGE_GEN 0.05
+
+#define EMBEDDED_RECOIL(x)     list(recoil_buildup = 1.4*x,  brace_penalty = 0.1*x,   one_hand_penalty = 0*x     )
```
[the remark this answers is not in the document]
Comment (Humonitarian) [reply code/__DEFINES/weapons.dm:79]: While I understand your point, the values in the define were meant for futureproofing. The only embedded gun is makeshift, and meant to be somewhat underpowered. If a proper one is added later, this define will be convenient.

[the remark this answers is not in the document]
Comment (SirRichardFrancis) [reply code/__DEFINES/weapons.dm:79]: Fair enough.

File: code/__DEFINES/weapons.dm [hunk 9]
```diff
@@ -75,3 +75,11 @@
...
+#define HANDGUN_RECOIL(x)      list(recoil_buildup = 1.2*x,  brace_penalty = 0.2*x,   one_hand_penalty = 0.6*x   )
+#define SMG_RECOIL(x)          list(recoil_buildup = 1.0*x,  brace_penalty = 0.3*x,   one_hand_penalty = 1.2*x   )
+#define CARBINE_RECOIL(x)      list(recoil_buildup = 0.8*x,  brace_penalty = 0.4*x,   one_hand_penalty = 1.8*x   )
+#define RIFLE_RECOIL(x)        list(recoil_buildup = 0.6*x,  brace_penalty = 0.5*x,   one_hand_penalty = 2.4*x   )
+#define LMG_RECOIL(x)          list(recoil_buildup = 0.4*x,  brace_penalty = 0.6*x,   one_hand_penalty = 3*x     )
+#define HMG_RECOIL(x)          list(recoil_buildup = 0.2*x,  brace_penalty = 0.7*x,   one_hand_penalty = 3.6*x   )
```
[the remark this answers is not in the document]
Comment (Humonitarian) [reply code/__DEFINES/weapons.dm:85]: Thx 👍

File: code/modules/mob/living/living_recoil.dm (hunk 3 above)
Comment (Humonitarian) [reply code/modules/mob/living/living_recoil.dm:10]: Thx for the reminder, I'll have to get around to this too

File: code/datums/recoil.dm [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+/datum/recoil
+	var/recoil_buildup
+	var/brace_penalty
+	var/one_hand_penalty
+
+/datum/recoil/New(recoil_buildup = 0, brace_penalty = 0, one_hand_penalty = 0)
+	src.recoil_buildup = recoil_buildup
+	src.brace_penalty = brace_penalty
+	src.one_hand_penalty = one_hand_penalty
+	tag = RECOILID
```
[the remark this answers is not in the document]
Comment (Humonitarian) [reply code/datums/recoil.dm:23]: What would be the purpose? Removing any problems that might come up with the vars of the same name guns have?

[the remark this answers is not in the document]
Comment (SirRichardFrancis) [reply code/datums/recoil.dm:23]: Contribution guidelines.

File: code/datums/recoil.dm [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+  return getRecoil((isnull(recoil_buildup) ? src.recoil_buildup : recoil_buildup),\
+				  (isnull(brace_penalty) ? src.brace_penalty : brace_penalty),\
+				  (isnull(one_hand_penalty) ? src.one_hand_penalty : one_hand_penalty))
+
+/datum/recoil/proc/getRating(rating)
+	return vars[rating]
+
+// Readable via recoil datum
+/datum/recoil/proc/getList()
+	return list(RECOIL_BASE = recoil_buildup, RECOIL_TWOHAND = brace_penalty, RECOIL_ONEHAND = one_hand_penalty)
```
[the remark this answers is not in the document]
Comment (Humonitarian) [reply code/datums/recoil.dm:41]: It's a proc with the same purpose as in datum/armor.dm, except there is no use for it yet. We could remove it, considering I don't think we'll need it anytime soon.

[the remark this answers is not in the document]
Comment (SirRichardFrancis) [reply code/datums/recoil.dm:41]: Doesn't really matter, just saying.

File: code/modules/projectiles/projectile/energy.dm (hunk 7 above)
Comment (Humonitarian) [reply code/modules/projectiles/projectile/energy.dm:102]: It's an unused bay leftover gun that's called a plasma gun despite having nothing to do with plasma guns.

File: code/modules/projectiles/projectile/beams.dm (hunk 6 above)
Comment (Humonitarian) [reply code/modules/projectiles/projectile/beams.dm:15]: Mathematically a thing, even if insignificant.

This amount of recoil is small enough to not be noticable, but enough to allow us to implement gunmods that can make it _very_ noticable. Thunder damage mod comes to mind.

File: code/modules/projectiles/guns/projectile/automatic/ak47.dm [hunk 12]
```diff
@@ -138,14 +137,14 @@
...
-		recoil_buildup = 1.5
 		w_class = ITEM_SIZE_BULKY
 		folded = FALSE
+		recoil = RIFLE_RECOIL(0.7)
 	else
 		to_chat(user, SPAN_NOTICE("You fold the stock on \the [src]."))
-		recoil_buildup = 1.8
 		w_class = ITEM_SIZE_NORMAL
 		folded = TRUE
+		recoil = CARBINE_RECOIL(0.7)
```
[the remark this answers is not in the document]
Comment (Humonitarian) [reply code/modules/projectiles/guns/projectile/automatic/ak47.dm]: Carbine recoil gives roughly 8% more recoil when wielded two-handed, 16% when braced.

**Comment by ValoTheValo** [comment]
okay i have tested this on the test server and found bugs and unmentioned balance changes
The Novakovic Rifle has broken recoil and gives a message like you are not wielding it even when wielded

"You struggle to keep SA BR .30 "Novakovic" (Wielded) on target while carrying it!" this also effects its recoil
the Takeshi had several unannounced nerfs. it now gets movement recoil even with no armor and starts with it already having some recoil. it also appears to be more inaccurate than before

i was unable to fire over reinforced tables. don't know if this is to do with this PR or just Eris Jank

ill do more testing soon

**Comment by SirRichardFrancis** [comment]
Firstly there is absolutely no reason for this to be made, secondly it will screw up recoil and make all these guns impossible to shoot, thirdly, one and possibly more gun recoil changes are direct violation to contributing guidelines, you should try reading them.
https://wiki.cev-eris.com/Content_GuidelinesEn

**Comment by ValoTheValo** [comment]
Futher update. its not only the Takeshi effected by recoil from movement, every single gun now has it regardless of stiffness of armor

**Comment by ValoTheValo** [comment]
> Firstly there is absolutely no reason for this to be made, secondly it will screw up recoil and make all these guns impossible to shoot, thirdly, one and possibly more gun recoil changes are direct violation to contributing guidelines, you should try reading them. https://wiki.cev-eris.com/Content_GuidelinesEn

that honestly sums of up my feelings yeah. even if it is a shitpost

**Comment by ValoTheValo** [comment]
Recoil appears to degrade at a much slower pace than before

**Comment by ValoTheValo** [comment]
apparently some of the LMG changes are just due to cursor shitcode

File: code/modules/mob/living/carbon/human/human_recoil.dm [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+	var/suit_stiffness = 0
+	var/uniform_stiffness = 0
+	if(H.wear_suit)
+		suit_stiffness = H.wear_suit.stiffness
+	if(H.w_uniform)
+		uniform_stiffness = H.w_uniform.stiffness
+
+	base_recoil += suit_stiffness + suit_stiffness * uniform_stiffness // Wearing it under actual armor, or anything too thick is extremely uncomfortable.
+
+	add_recoil(base_recoil)
```
Comment (hyperioo) [inline code/modules/mob/living/carbon/human/human_recoil.dm:16]:
```suggestion
	add_recoil(base_recoil)

```

**Review by hyperioo — APPROVED** [review]

**Pull request merged**